File: algorithms/C++/BinarySearch/BinarySearch_Recursive.c
#include <stdio.h>
void print(int *arr, int s, int e)
{
    printf( "Printing Array\n" );
    for (int i = 0; i <=e; i++)
    {
        printf("%d ",arr[i]);
    }
    printf("\n");
}
int binarySearch(int *arr, int s, int e, int key)
{
    

    if (s > e)
    {
        return 0;
    }
    int mid = s + (e - s) / 2;

    if (arr[mid] == key)
    {
        return 1;
    }
   
    if (arr[mid] < key)
    {
        return binarySearch(arr, mid + 1, e, key);
    }
     else
    {
        return binarySearch(arr, s, mid - 1, key);
    }
}
int main()
{
    int n;
    printf("Enter size of array: ");
    scanf("%d",&n);
    int arr[n];
    printf("Enter element into the array");
    for(int i = 0; i < n; i++)
    {
      scanf("%d",&arr[i]);
    }
    int size = n;
    int key;
    printf("Enter the value you want to search: ");
    scanf("%d",&key);
    int ans = binarySearch(arr, 0, size - 1, key);
    printf("Present or not %d",ans);
    
    return 0;
}

File: algorithms/searching/binary-search/kotlin/BinarySearchRecursive.kt
fun IntArray.binarySearch(lowerIndex: Int, upperIndex: Int, x: Int): Int {
    val arr = this
    if (upperIndex >= lowerIndex) {
        val mid = lowerIndex + (upperIndex - lowerIndex) / 2

        // If the element is found
        if (arr[mid] == x)
            return mid

        /* If the element is smaller than mid, then it can only be present in left subarray
        else the element can only be present in right subarray */
        return if (arr[mid] > x) binarySearch(lowerIndex, mid - 1, x)
        else binarySearch(mid + 1, upperIndex, x)
    }
    // We reach here only when the element is not present in array
    return -1
}

fun main(args: Array<String>) {
    println("Enter the size of array:")
    val n = readLine()!!.toInt()
    println("Enter the elements of array:")
    var arr = IntArray(n) { readLine()!!.toInt() }
    println("Enter a number to search:")
    val x = readLine()!!.toInt()
    arr = arr.sortedArray() // Sorting the array in ascending order
    println("Array in ascending order: ")
    for (l in arr)
        print("$l ")
    println()
    val result = arr.binarySearch(0, n - 1, x)
    if (result == -1)
        println("Element not present")
    else
        println("Element found at index $result")
}
